### ISSUE TITLE
Unlock your wallet in an asynchronous procedure

### ISSUE THREAD

**Issue by Senyoret1** [skycoin/skycoin-web#162]
After entering the seed in the unlock wallet modal window and pressing the "Unlock" button, the browser freezes, which is not good for the UX. To avoid this, the wallet must be unlock in an asynchronous process and the "Unlock" button should show the loading animation.

**Comment by alexandermushinskiy** [comment]
@Senyoret1, we added timeout inside `walletService.unlockWallet()` to increase the "unlocking time" and the UI was not frozen, we can click outside the modal window to close it and locate to other pages during "unlocking process".
Would it be enough just to add loading animation in scope of this issue?

**Comment by Senyoret1** [comment]
The problem is a little more complex than it seems. I'm not sure what the timeout did, but try placing this code inside `WalletService.unlockWallet()`:

```
for (let i=0; i<5000000000; i++) {
    let tmp = 500 / 2 * 2 /2 * 2;
}
```

With that code it is much more evident how the UI suffers when unlocking a wallet.

Without that extra code, unlocking a wallet with one address with a Pentum anniversary edition takes approximately 1.9 seconds. Because of how the code works, a wallet with 3 addresses would take 5.7 seconds. With a faster processor the UI suffer less, but many users will use modest pcs.

**Comment by Senyoret1** [comment]
The 1.9 seconds delay was measured usign Chrome. I tested a little more and found that in Firefox it takes 3.5 seconds and in Edge more than 5.

**Comment by alexandermushinskiy** [comment]
@Senyoret1, we tried to use promises and observable and it doesn't help to execute unlock functionality asynchronously. We changed `wallet.addresses.forEach(address=> ..` to recursion but it also doesn't help. In all cases the UI is blocked. Some of our colleagues suggested us to use web workers for this case but it makes solution as complicated. Do you have any ideas?
As an easy and quick fix we can add a timeout inside `walletService.unlockWallet()` and user will see the loading animation on a button during the execution.

**Comment by Senyoret1** [comment]
I remember that I did a quick test with timeout and it gave problems, but at this moment I do not remember well what went wrong. I have not looked very closely at what the solution would be using web workers, but it is more likely that it will be necessary to use them. Seeing this, please give me a little time to do some testing to see if there are other options and try to define what would be the best approach.

**Comment by Senyoret1** [comment]
@alexandermushinskiy https://github.com/skycoin/skycoin-web/pull/264 adds some code to execute procedures in web workers without complicating much the code. Please take a look at it, although I think this may interfere with the process of adapting the cipher test suite, so it would probably be good to wait for those tests to be completed.

**Comment by alexandermushinskiy** [comment]
Yes, we saw that PR, it looks really good. We have already tested your changes with unlock wallet process locally and it works as expected.

**Comment by Senyoret1** [comment]
Good, the PR is merged.

**Comment by alexandermushinskiy** [comment]
The issue is implemented.

Have the PR settings changed? By default it wants to be merge to develop branch (usually we merged to master) and travis-ci is not started (required some expectation).

**Comment by Senyoret1** [comment]
Sorry for not notifying you before. Now the main branch is develop, master will be used for stable versions. The solution to the problem with Travis is in https://github.com/skycoin/skycoin-web/pull/273, it just need one review to merge it.

**Comment by alexandermushinskiy** [comment]
@Senyoret1, please be informed that your review comment is fixed.
Sorry for so many commits, I had always conflicts.
In my personal opinion generating one address by web worker takes too long (you can generate new address).

**Comment by Senyoret1** [comment]
@alexandermushinskiy I reviewed the PR and now I understand why you wanted to add a web worker for creating several addresses at the same time. In my tests, creating an address without the web worker takes something like 2 seconds, whereas with the web worker it takes around 10, so creating 2 addresses would take 4 seconds without the web worker and 20 seconds with it, which is a significant difference. A web worker able to create several addresses at the same time would decrease the time to 12 seconds, but there is a another option.

The huge difference in time is because each time the web worker is executed, `importScripts('../main.js')` is called, an instruction that is very time consuming. However, if the web worker is reused, only the first execution would take more than 2 seconds, the following executions would be almost as fast as creating the address outside the web worker, so it would not be necessary to have a web worker able to create more than one address at a time. Also, if the web worker is initialized as soon as the wallet is opened, then the first execution would also be quick, so creating 2 addresses would take 4 seconds, instead of 20 or 12.

I did some tests to see if that works, so the majority of the code is already writen and I should be adding it soon. After that, the problem of the excessive slowness should be fixed.
